### ISSUE TITLE
AdminClient hangs on dispose

### ISSUE THREAD

**Issue by HenningRoigaard** [confluentinc/confluent-kafka-dotnet#671]
The AdminClient hangs on dispose

The repro scenario is simple using a dotnet core app:

```
static async Task Main(string[] args)
{
      using (var adminClient = new AdminClient(new AdminClientConfig() { BootstrapServers = "localhost:9092" }))
      {
        var topicSpecification = new TopicSpecification() { Name = Guid.NewGuid().ToString(), NumPartitions = 1, ReplicationFactor = 1 };
        await adminClient.CreateTopicsAsync(new[] { topicSpecification });
      } // Hangs here
}
```
The issue disappears if I stop using _await_ and use _Wait()_ instead, i.e. 
```
static async Task Main(string[] args)
{
      using (var adminClient = new AdminClient(new AdminClientConfig() { BootstrapServers = "localhost:9092" }))
      {
        var topicSpecification = new TopicSpecification() { Name = Guid.NewGuid().ToString(), NumPartitions = 1, ReplicationFactor = 1 };
        adminClient.CreateTopicsAsync(new[] { topicSpecification }).Wait();
      }
}
```
Due  to #670 I cannot get debug output to supplement this issue.

Using Visual Studios Task Window I see a deadlock which seems like a likely candidate

![visualstudio_adminclientdispose_screenshot](https://user-images.githubusercontent.com/25740721/48059925-c3cda100-e1ba-11e8-930c-f277467ee302.png)

Checklist
=========
Please provide the following information:

 - [x] Confluent.Kafka nuget version: 1.0.0-beta2
 - [x] Apache Kafka version: cp-kafka 3.3.0
 - [x] Client configuration: See code example
 - [x] Operating system: Windows 10
 - [ ] Provide logs (with "debug" : "..." as necessary in configuration)
 - [ ] Provide broker log excerpts
 - [ ] Critical issue

**Comment by LasseNisted** [comment]
A bit of debugging shows that the order of events seems to be:

1. The AdminClient construction the StartPollTask is started with Task.Factory.StartNew and set as callbackTask.
2. Main invokes CreateTopicsAsync and awaits the retrieved task which comes from a TaskCompletionSource - this becomes a continuation.
3. When the poll retrieves the result from Kafka it sets it on the TaskCompletionSource via TrySetResult
4. The TrySetResult synchronously invokes continuations and thereby invokes the rest of Main
5. Main invokes Dispose which waits for callbackTask to finish. But callbackTask is the one invoking main through TrySetResult. It deadlocks.

The following is a quick reproduction to illustrate the issue without Kafka involvement:
```
class Program
{
    static TaskCompletionSource<byte> taskCompletionSource = new TaskCompletionSource<byte>(/* TaskCreationOptions.RunContinuationsAsynchronously */);

    static Task StartPollTask(CancellationToken ct)
            => Task.Factory.StartNew(() =>
            {
              while (!ct.IsCancellationRequested)
              {
                Thread.Sleep(500);
                taskCompletionSource.TrySetResult(1);
              }
            }, ct, TaskCreationOptions.LongRunning, TaskScheduler.Default);

    static async Task Main(string[] args)
    {
      var cancellationTokenSource = new CancellationTokenSource();

      var callbackTask = StartPollTask(cancellationTokenSource.Token);

      var dummy = await taskCompletionSource.Task;

      cancellationTokenSource.Cancel();
      callbackTask.Wait();
      cancellationTokenSource.Dispose();
    }
  }
```

It may be a bit brittle to invoke continuations synchronously from inside the poll loop through TrySetResult. A solution would be to construct the TaskCompletionSource with the TaskCreationOptions.RunContinuationsAsynchronously argument.

**Comment by alexandery** [comment]
This is still the issue with the latest (released 1.0.0) version. Awaiting topics creation results in the executable simply exiting with no errors and not exceptions. Admin client isn't even being disposed at the time of the call.

I think this should really be looked into.

**Comment by mhowlett** [comment]
resolved in 1.0

**Comment by teemka** [comment]
Hi,
I found this issue after experiencing similar problem with application deadlocking on `Producer.Dispose()`.

The problem is similar - I used `TaskCompletionSource` in [Producer.Produce](https://github.com/confluentinc/confluent-kafka-dotnet/blob/29995e6b48fc914eeb43e0c8f62665ca5c87e9af/src/Confluent.Kafka/Producer.cs#L855) method. The delivery handler called the TCS and the task continued to dispose the producer which resulted in circular dependency (deadlock)

The Producer issue is still present in Confluent.Kafka 2.8.0
